File: apps/backend/app/models/invoice.py
"""
Invoice model - represents electronic invoices (comprobantes electrónicos SUNAT).
"""

from sqlalchemy import Column, DateTime, Float, ForeignKey, Index, Integer, JSON, String, UniqueConstraint
from sqlalchemy.orm import relationship

from app.models.base import Base, TimestampMixin


class Invoice(Base, TimestampMixin):
    """
    Invoice model - represents electronic invoices for SUNAT (Facturas, Boletas, NC, ND).

    Types:
    - 01: Factura
    - 03: Boleta
    - 07: Nota de Crédito
    - 08: Nota de Débito
    """

    __tablename__ = "invoices"

    # Multitenant
    tenant_id = Column(
        Integer,
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Tenant (company) this invoice belongs to",
    )
    tenant = relationship("Tenant", back_populates="invoices")

    # Order reference
    order_id = Column(
        Integer,
        ForeignKey("orders.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Order this invoice is for",
    )
    order = relationship("Order", back_populates="invoices")

    # Invoice identification
    invoice_type = Column(
        String(2),
        nullable=False,
        comment="01=Factura, 03=Boleta, 07=NC, 08=ND",
    )
    serie = Column(
        String(4),
        nullable=False,
        comment="Serie del comprobante (ej: F001, B001)",
    )
    correlativo = Column(
        Integer,
        nullable=False,
        comment="Número correlativo del comprobante",
    )

    # Issuer (emisor)
    emisor_ruc = Column(
        String(11),
        nullable=False,
        comment="RUC del tenant emisor",
    )
    emisor_razon_social = Column(
        String(200),
        nullable=False,
        comment="Razón social del tenant",
    )

    # Customer (cliente)
    cliente_tipo_documento = Column(
        String(1),
        nullable=False,
        comment="1=DNI, 6=RUC",
    )
    cliente_numero_documento = Column(
        String(11),
        nullable=False,
        comment="DNI o RUC del cliente",
    )
    cliente_razon_social = Column(
        String(200),
        nullable=False,
        comment="Nombre o razón social del cliente",
    )

    # Amounts
    currency = Column(
        String(3),
        default="PEN",
        nullable=False,
        comment="PEN, USD",
    )
    subtotal = Column(
        Float,
        nullable=False,
        comment="Subtotal sin IGV",
    )
    igv = Column(
        Float,
        nullable=False,
        comment="IGV (18%)",
    )
    total = Column(
        Float,
        nullable=False,
        comment="Total a pagar",
    )

    # Line items
    items = Column(
        JSON,
        nullable=False,
        comment="Line items desde order.line_items",
    )

    # Reference for NC/ND (auto-reference)
    reference_invoice_id = Column(
        Integer,
        ForeignKey("invoices.id", ondelete="SET NULL"),
        nullable=True,
        comment="Invoice referenciado para NC/ND",
    )
    reference_invoice = relationship(
        "Invoice",
        remote_side="Invoice.id",
        foreign_keys=[reference_invoice_id],
        backref="credit_debit_notes",
    )
    reference_type = Column(
        String(2),
        nullable=True,
        comment="Tipo de documento referenciado",
    )
    reference_serie = Column(
        String(4),
        nullable=True,
        comment="Serie del documento referenciado",
    )
    reference_correlativo = Column(
        Integer,
        nullable=True,
        comment="Correlativo del documento referenciado",
    )
    reference_reason = Column(
        String(200),
        nullable=True,
        comment="Motivo de la NC/ND",
    )

    # eFact integration
    efact_ticket = Column(
        String(100),
        unique=True,
        index=True,
        nullable=True,
        comment="UUID de eFact",
    )
    efact_status = Column(
        String(20),
        default="pending",
        index=True,
        nullable=False,
        comment="pending, processing, success, error",
    )
    efact_response = Column(
        JSON,
        nullable=True,
        comment="CDR de SUNAT",
    )
    efact_error = Column(
        String(500),
        nullable=True,
        comment="Mensaje de error si falla",
    )
    efact_sent_at = Column(
        DateTime,
        nullable=True,
        comment="Timestamp cuando se envió a eFact",
    )
    efact_processed_at = Column(
        DateTime,
        nullable=True,
        comment="Timestamp cuando eFact procesó el comprobante",
    )

    # Constraints and indexes
    __table_args__ = (
        UniqueConstraint(
            "tenant_id",
            "serie",
            "correlativo",
            name="uq_tenant_serie_correlativo",
        ),
        Index("ix_invoices_order_invoice_type", "order_id", "invoice_type"),
    )

    def __repr__(self) -> str:
        """String representation of Invoice."""
        return (
            f"<Invoice(id={self.id}, type={self.invoice_type}, "
            f"serie={self.serie}, correlativo={self.correlativo}, "
            f"total={self.total}, status={self.efact_status})>"
        )

File: apps/frontend/lib/types/invoice.ts
/**
 * Invoice Types - Tipos para el sistema de facturación electrónica
 * Integración con eFact-OSE para SUNAT (Perú)
 */

/**
 * LineItem - Item de línea en una factura
 * Representa un producto/servicio facturado
 */
export interface LineItem {
  descripcion: string;
  cantidad: number;
  precio_unitario: number;
  subtotal: number;
  igv: number;
  total: number;
}

/**
 * Invoice - Comprobante electrónico (Factura, Boleta, NC, ND)
 * Incluye todos los campos del backend para facturación SUNAT
 */
export interface Invoice {
  // IDs
  id: number;
  tenant_id: number;
  order_id: number;

  // Tipo de comprobante
  invoice_type: "01" | "03" | "07" | "08"; // 01=Factura, 03=Boleta, 07=NC, 08=ND

  // Numeración
  serie: string; // Ej: "F001", "B001", "NC01", "ND01"
  correlativo: number; // Número secuencial
  full_number: string; // Computed: "F001-00000123"

  // Emisor (Tenant/Empresa)
  emisor_ruc: string;
  emisor_razon_social: string;

  // Cliente
  cliente_tipo_documento: string; // "1"=DNI, "6"=RUC, "4"=CE, etc.
  cliente_numero_documento: string;
  cliente_razon_social: string;

  // Moneda y montos
  currency: "PEN" | "USD";
  subtotal: number;
  igv: number; // Impuesto General a las Ventas (18%)
  total: number;

  // Items/Líneas del comprobante
  items: LineItem[];

  // Campos de referencia (solo para NC/ND)
  reference_invoice_id?: number; // ID de la factura referenciada
  reference_type?: string; // Tipo de referencia
  reference_serie?: string; // Serie del comprobante referenciado
  reference_correlativo?: number; // Correlativo del comprobante referenciado
  reference_reason?: string; // Motivo de la NC/ND

  // Estado y respuesta de eFact-OSE
  efact_status: 'pending' | 'processing' | 'success' | 'error';
  efact_ticket?: string; // Ticket de SUNAT
  efact_response?: any; // Respuesta completa de SUNAT
  efact_error?: string; // Mensaje de error si falla
  efact_sent_at?: string; // Fecha de envío a SUNAT
  efact_processed_at?: string; // Fecha de procesamiento por SUNAT

  // Timestamps
  created_at: string;
  updated_at: string;
}

/**
 * InvoiceCreate - Datos para crear un nuevo comprobante
 * Usado al generar factura desde una orden
 */
export interface InvoiceCreate {
  invoice_type: "01" | "03" | "07" | "08"; // Tipo de comprobante
  serie: string; // Serie configurada (ej: "F001", "B001")
  reference_invoice_id?: number; // Solo para NC/ND - ID del comprobante a referenciar
  reference_reason?: string; // Solo para NC/ND - Motivo de la nota
  // Customer data overrides (optional)
  cliente_tipo_documento?: string; // Override document type
  cliente_numero_documento?: string; // Override document number
  cliente_razon_social?: string; // Override customer name
  cliente_email?: string; // Override customer email
}

/**
 * InvoiceSerie - Serie de numeración para comprobantes
 * Cada serie tiene su propio contador de correlativo
 */
export interface InvoiceSerie {
  id: number;
  tenant_id: number;
  invoice_type: "01" | "03" | "07" | "08";
  serie: string; // 4 caracteres alfanuméricos
  last_correlativo: number; // Último correlativo usado (siguiente = last + 1)
  is_active: boolean;
  description?: string;
  created_at: string;
  updated_at: string;
}

/**
 * InvoiceSerieCreate - Datos para crear una nueva serie
 */
export interface InvoiceSerieCreate {
  invoice_type: "01" | "03" | "07" | "08";
  serie: string; // 4 caracteres (ej: "F001", "B001")
  description?: string;
  is_active?: boolean;
}

/**
 * InvoiceListResponse - Respuesta paginada de lista de comprobantes
 */
export interface InvoiceListResponse {
  total: number;
  items: Invoice[];
  skip: number;
  limit: number;
}

/**
 * INVOICE_TYPES - Constantes de tipos de comprobantes SUNAT
 */
export const INVOICE_TYPES = {
  FACTURA: '01',
  BOLETA: '03',
  NOTA_CREDITO: '07',
  NOTA_DEBITO: '08',
} as const;

/**
 * INVOICE_TYPE_NAMES - Mapeo de códigos a nombres legibles
 * Según catálogo SUNAT de tipos de comprobantes
 */
export const INVOICE_TYPE_NAMES: Record<string, string> = {
  '01': 'Factura Electrónica',
  '03': 'Boleta de Venta Electrónica',
  '07': 'Nota de Crédito Electrónica',
  '08': 'Nota de Débito Electrónica',
};

/**
 * INVOICE_STATUS_NAMES - Mapeo de estados a textos
 * Estados del proceso de envío a SUNAT
 */
export const INVOICE_STATUS_NAMES: Record<string, string> = {
  pending: 'Pendiente',
  processing: 'Procesando',
  success: 'Validado por SUNAT',
  error: 'Error en SUNAT',
};

/**
 * INVOICE_STATUS_COLORS - Colores Tailwind para estados
 */
export const INVOICE_STATUS_COLORS: Record<string, string> = {
  pending: 'bg-gray-50 text-gray-700 border-gray-200',
  processing: 'bg-yellow-50 text-yellow-700 border-yellow-200',
  success: 'bg-green-50 text-green-700 border-green-200',
  error: 'bg-red-50 text-red-700 border-red-200',
};

// Mantener LABELS por compatibilidad con código existente
export const INVOICE_TYPE_LABELS = INVOICE_TYPE_NAMES;
export const INVOICE_STATUS_LABELS = INVOICE_STATUS_NAMES;
